### ISSUE TITLE
Add lossless optimization for JPEG images embedded to the PDF

### ISSUE THREAD

**Issue by pts** [pts/pdfsizeopt#41]
1. EXIF tags etc. can be removed by Python code, see *Removing JPEG metadata (e.g. comments, JFIF, Exif etc.) manually* in info.txt for details. Don't use `jpegtran -copy none`, because it would keep some unnecessary metadata.

2. Smaller Huffman tables can be generated. `jpegtran -optimize` can do this. mozjpeg's jpegtran doesn't create any smaller files. This is equivalent to the lossless mode of `jpegoptim` and `imgopt`.

3. [jpgcrush](https://github.com/wafflesnatcha/bin/blob/master/jpgcrush) and [jpegrescan](https://raw.githubusercontent.com/kud/jpegrescan/master/jpegrescan) cannot be used, because they create progressive JPEG output, which PDF doesn't support.

FYI If mozjpeg's jpegtran is used, then it should be invoke with `-revert`, otherwise it enables `-progressive` by default, and PDF doesn't support progressive JPEG.

**Comment by rbrito** [comment]
Hi, Péter.

On Sep 15 2017, Péter Szabó wrote:
> 1. EXIF tags etc. can be removed by Python code, see *Removing JPEG
> metadata (e.g. comments, JFIF, Exif etc.) manually* in info.txt for
> details. Don't use `jpegtran -copy none`, because it would keep some
> unnecessary metadata.

Nice to know and highly useful for "anonymization" (or, at least, one step
further towards it) of PDF files. Another option is to use `jhead -purejpg`,
but using fewer dependencies is, of course, better.

> 2. Smaller Huffman tables can be generated. `jpegtran -optimize` can do
> this. mozjpeg's jpegtran doesn't create any smaller files. This is
> equivalent to the lossless mode of `jpegoptim` and `imgopt`.

jpegran is widely available on Linux distributons and I suspect that it can
be made small even if you statically link it. Didn't know that mozjpeg's
jpegtran didn't create smaller files wrt to Huffman tables. I assumed that
it did.

> 3. [jpgcrush](https://github.com/wafflesnatcha/bin/blob/master/jpgcrush)
> and
> [jpegrescan](https://raw.githubusercontent.com/kud/jpegrescan/master/jpegrescan)
> cannot be used, because they create progressive JPEG output, which PDF
> doesn't support.

Didn't know that.

Anyway, optimizing the JPEG files is awesome for the future!


Thanks once again,

-- 
Rogério Brito : rbrito@{ime.usp.br,gmail.com} : GPG key 4096R/BCFCAAAA
http://cynic.cc/blog/ : github.com/rbrito : profiles.google.com/rbrito
DebianQA: http://qa.debian.org/developer.php?login=rbrito%40ime.usp.br

.

**Comment by rbrito** [comment]
Is this too hard to implement? It would rock to have this, since there are many files that have JPEGs embedded that could use some optimization of the kind of jpegtran...

Regarding jpgcrush & co., I don't think that it would be a problem to upgrade PDFs to version 1.3... Almost every PDF out there is with a version later than this...

**Comment by pts** [comment]
> Is this too hard to implement?

No, it's relatively straightforward, but it needs time to implement, which I don't have immediately. (Funding could make a difference here.) The first step would be running jpgcrush, jpegrescan and mozjpeg's `jpegtran -optimize` on many JPEG files in PDFs, and figuring out if there is a clear winner (i.e. smallest output size). If there is a winner, it should be compiled and  added to pdfsizeopt_libexec, and then the calling code should be added to pdfsizeopt. The removing of metadata is also not hard to implement. It's about 8 hours of work in total.

**Comment by rbubley** [comment]
If you ever get a chance to look at this, another optimizer worth looking at is  [jpegoptim](https://github.com/tjko/jpegoptim).

**Comment by rbrito** [comment]
Just for the record, I [packaged jpgcrush][jpgcrush] for Debian-based systems (properly patched to be Debian-friendly with files not under `/usr/local` etc.).

With it installed, I can use [a script of mine][optimize_jpegs] to call `jpgcrush` on some JPEG files (depending on the colorspace---I know how to handle RGB and some ICC-based JPEGs).

[jpgcrush]: https://launchpad.net/~rbrito/+archive/ubuntu/ppa
[optimize_jpegs]: https://github.com/rbrito/scripts/blob/master/optimize_jpegs.py

**Comment by pts** [comment]
FYI New version of the script is available at: https://github.com/rbrito/scripts/blob/master/optimize_pdfs.py

**Comment by pts** [comment]
FYI The design is the following.

For each image object with its last filter being `/DCTDecode`: uncompress everything exept for `/DCTDecode`.

For each image object with `/Filter /DCTDecode`:

1. Parse the JPEG bitstream (i.e. the `stream` data of the PDF image object), find all markers until the JPEG image data itself.
2. If the JFIF marker is missing, display a warning, give up, and keep the image object unchanged.
3. If the JPEG dimensions and number of components don't match the image dimensions, display a warning, give up, and keep the image object unchanged.
4. If the Adobe APP14 (APPE) marker is peresent (this is the one with `/ColorTransform`), save its contents for later use.
5. Remove all APP* and COM markers (including JFIF). This removes all metadata (including Exif, XMP, IPTC and ICC).
6. If a SOF1, SOF3 ... SOF15 marker (any) is detected, display a warning, give up, and keep the image object unchanged.
7. If an unknown marker is detected (i.e. not on the whitelist), display a warning, give up, and keep the image object unchanged.
8. Add a JFIF marker without thumbnail (18 bytes). Add back the Adobe APP14 marker (if it was present).
9. Save the the result to *source.jpg*.
10. Run a bunch of inernal and external JPEG image optimizers on *source.jpg*, creating files *opt1.jpg* ... *optN.jpg*.
11. Read each *optK.jpg*, remove all APP* and COM markers (including JFIF), add a JFIF marker without thumbnail (18 bytes), add back the Adobe APP14 marker (if it was present).
12. Choose the smallest of *source.jpg* and all *optK.jpg*, replace the stream of the PDF image object with it.
13. If `/ColorTransform` is specified in the image object, and it either has the default value, or the Adobe APP14 marker was also present, drop the `/ColorTransform`.
14. Bump the PDF version to 1.3 if the chosen stream is a progressive JPEG (SOF2 marker) rather than baseline JPEG (SOF0 marker).

JPEG image optimizers included by default:

* *jpegtran* of [mozjpeg](https://libjpeg-turbo.org/About/Mozjpeg): it optimizes much more by default than traditional *jpegtran*
* *jpgcrush* and its included *jpegrescan*, both of which call the traditional *jpegtran*. It makes some images a 100 bytes smaller than *jpegtran* of mozjpeg. The logic will be ported from Perl to Python. Traditional *jpegtran* will be eumulated with `jpegtran -revert` in mozjpeg. Extra testing will be done to make sure it works with 4 components (e.g. `/Colorspace /DeviceCMYK`).

There will be a `--use-jpeg-optimizer=...` command-line flag (similar to `--use-image-optimizer`, maybe they will be merged) to specify which JPEG image optimizers to try. If will be possible to use other optimizers (as external programs) as well, in addition those included.

**Comment by zvezdochiot** [comment]
@pts say:
>jpegtran

jpegtran only makes sense if the task is to split one object into several smaller ones. To control the size/quality of an object, [jpegoptim](https://github.com/tjko/jpegoptim) should be used.

**Comment by pts** [comment]
@zvezdochiot, I've extended my comment with some info about how *jpegoptim* can be used by *pdfsizeopt* in the future.

**Comment by zvezdochiot** [comment]
@pts say:
>for lossless JPEG image optimization

[jpegoptim](https://github.com/tjko/jpegoptim) not only lossless. See option `-m`. ;)

**Comment by pts** [comment]
> [jpegoptim](https://github.com/tjko/jpegoptim) not only lossless. See option `-m`. ;)

True, there is no confusion about this fact. My comment https://github.com/pts/pdfsizeopt/issues/41#issuecomment-1445066996 also indicates that *jpegoptim* is not only lossless.

This issue is about lossless JPEG image optimization only, so the lossy features of *jpegoptim* shouldn't be used here. See https://github.com/pts/pdfsizeopt/issues/95 for lossy JPEG image optimization in *pdfsizeopt*.

**Comment by pts** [comment]
@zvezdochiot, do you have some JPEG files to share for which *jpegoptim* (`jpegoptim -s --all-progressive`) generates very small output? I suspect that *jpegtran* of mozjpeg will generate smaller output, but we should try and compare

**Comment by zvezdochiot** [comment]
@pts say:
>do you have some JPEG files to share for which jpegoptim

Not. No. I am now far from this topic. I'm working on automating PDF splitting into objects using `csplit` and replacing entire `/DCT` objects with `/CCITT` (use `LibTIFF`) and `/Flate` (use `sam2p`).

**Comment by Adreitz** [comment]
Just a couple other utilities to mention related to lossless JPEG optimization. It's very slow because it is brute force, but jpegultrascan (https://github.com/MegaByte/jpegultrascan) seems to produce the best results for me. I've used it simultaneously calling the IJG and mozjpeg versions of jpegtran. There is also pingo (https://css-ig.net/pingo.php), which is mostly a PNG optimizer but also does some limited, but very fast, JPEG optimization. Pingo is a great bang-for-buck utility, but it is Windows-only and closed-source.

**Comment by pts** [comment]
@Adreitz, thank you for suggesting *jpegultrascan* and *pingo*.

According to my quick measurements, *jpegultrascan* is about 1450 times slower than *jpegtran* of mozjpeg for a ~100 KiB JPEG input file, and makes the output JPEG file 0.06% smaller than *jpegtran* of mozjpeg. Thus it won't be included with *pdfsizeopt*.

Because it's a Windows-only tool, *pingo* won't be included either. Also, *pingo.exe* is huge (almost 3 MiB) considering what it can do.

However, it will be possible to use them (any many others) with `--use-jpeg-optimizer=...`.

**Comment by pts** [comment]
I was wondering which version of *jpegtran* to use. I've tried version 3.1 (2015-05-19) and version 4.1.1 (2022-09-15). The latter generated JPEG output files a few percent larger than the former. I've reported https://github.com/mozilla/mozjpeg/issues/433 to get advice.

**Comment by pts** [comment]
I've created a source port of *jpegtran* of mozjpeg on https://github.com/pts/pts-mozjpegtran. The executable programs used by *pdfsizeopt* will be built from these sources, cross-compiled on Linux.

**Comment by GitHubRulesOK** [comment]
Hmmm
 I come from a long time ago when the need for compression was, very high cost of media and handling, thus compression was an essential necessity for storage and transmission speeds.
 However the purpose of a modern PDF is to be readable by many as fast as possible, thus size of storage is no longer a necessity, its bangs per buck that is achieved by decompression/rendering time that is important in this modern century. 
 A few bytes saved by one authors exotic compressions (such as internet archive seem to use?) is magnified as a global warming delay, when each of thousands of readers take, extra seconds and CPU power to decompress for print or screen render.
There should be a balancing point where simpler decompression times outweighs a few KB saved.

**Comment by Adreitz** [comment]
@GitHubRulesOK This is not "exotic compression". The compression formats relevant to PDFs (JBIG2, JPEG, GIF, PNG, ZIP) are all well-known and asymmetric. The decompression time is not affected in an appreciable way by the compression time used. While there is a time and energy cost associated with optimized compression, it can easily be made up by the use of the optimized file. It is quicker to download and therefore uses less energy through network activity. It could even be marginally quicker to open if stored on a fairly slow medium. And, though the storage space recovered from optimizing a single file may not be very meaningful, it adds up, and I don't think anyone would complain if they found their drive to suddenly have 10-30% greater capacity.

Also, the JPEG optimizations discussed in this issue are generally very quick, as opposed to many PNG optimizations.

**Comment by jwillikers** [comment]
Is [jpegli](https://github.com/google/jpegli/) worth looking into? It's an improved JPEG encoder/decoder that backports applicable new techniques of JPEG XL to JPEG. It seems to be more efficient and just as fast as `mozjpeg`, although it seems to require using `djpegli` to decode the image and running `cjpegli` to re-encode the image to optimize it.

Nevermind, turns out it can't do lossless recompression: https://github.com/google/jpegli/issues/10#issuecomment-2221735156

**Comment by ocococococ** [comment]
Some very encouraging experiments with [qpdf](https://qpdf.readthedocs.io/en/stable/json.html) JSON options & [jpegoptim](https://www.kokkonen.net/tjko/src/man/jpegoptim.txt) as a first pass before using pdfsizeopt:

### Extract objects from input PDF as objects-* + qpdf.json files
`qpdf input.pdf
  --json
  --json-stream-data=file
  --json-stream-prefix=objects
  qpdf.json
`
### Lossless optimization of jpeg objects-* files (just ignore errors with non-jpeg files)
`jpegoptim objects-*`

### Recombine optimized objects with input PDF to produce temp PDF with optimized jpeg images
https://qpdf.readthedocs.io/en/stable/cli.html#optimizing-file-size

`QPDF_ZOPFLI=1 qpdf input.pdf
  --update-from-json=qpdf.json
  --compress-streams=y
  --decode-level=generalized
  --object-streams=generate
  --recompress-flate
  --compression-level=9
  --linearize
  temp.pdf`

### Final pass with **pdfsizeopt** to produce optimized output PDF
`pdfsizeopt temp.pdf output.pdf`

### Cleanup temporary files
`rm qpdf.json`
`rm objects-*`
`rm temp.pdf`
